### PR TITLE
MH-12778 Lock editor to avoid concurrent edits

### PR DESCRIPTION
**Pull request by ts23**
This PR creates a MediaPackageLock that avoids concurrent editing of a mediapackage. The mediapackage is locked with its id and the browsers sessionId when the user enters the editor. The lock is released when the user leaves the editor. If in case of a browser crash the lock is not released it will timeout automatically. The timeout is set to the duration of the video or 30 minutes (whichever is longer). 
The lock is not protecting against the same browser accessing the same video twice. 

This PR also introduces a protection against edits after a video has been published and an auto-save after 30 mins.

### CODE REVIEW

**Comment by slampunk** [comment]
Hi @ts23 

If you want to cover the case of the same browser opening the same video twice, you could look into using the Storage service. I've put together a gist for you:
https://gist.github.com/slampunk/89601db5d37e29bcedab9a2e679db686

I believe the Storage service is built atop the localStorage API. So there's the possibility for inter-tab communication to check whether sibling tabs are editing the same video too. In fact, inter-tab communication would be useful in the case where the browser has crashed before (as a result, not clearing the lock), and the new tab needs to find out if there is indeed an active edit taking place in a sibling tab at the current time.

Also, how do we review on github? I can review this PR.

**Comment by ts23** [comment]
The case of the same browser opening twice is intentionally not covered. One of the problems is that all rest calls are made twice (so the same browser needs to be allowed otherwise the editor would always lock itself automatically ... been there - done that

**Comment by staubesv** [comment]
Hi @ts23 I'm not asking you to implement this, but in case you would have time... It would be quite desirable to know which user currently holds the lock (Firstname, Lastname, E-mail address). Use case: User Alice want to edit the video, but Opencast lets her know that the video is currently locked. 20 minutes later, she tries again. Still the same. An hour later, she tries again (maybe this time, the lock is held by another person). From point of view of Alice, it won't be clear what is happening. If she knew who holds the lock, things might be more clear.

**Comment by ts23** [comment]
I see the point for this feature, but we would need it configurable as in our case all users edit as the same user - so this would not be useful for us.

**Comment by staubesv** [comment]
> [...] as in our case all users edit as the same user [...]

Seems there is some work ahead of you ;-)

Anyway, just had a look at the code - adding information about the user will be quite simple and as the Admin UI facade is an internal API, changing it later should not be a problem.

I'll do a follow-up PR once I have some time (i.e. won't be soon).

**Comment by ts23** [comment]
This "editing as same user" is intentional as we are using a  cas-front to access and then the user gets only access to the edit-ui

File: modules/admin-ui/src/main/resources/OSGI-INF/mediapackage_lock.xml [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<scr:component xmlns:scr="http://www.osgi.org/xmlns/scr/v1.1.0"
+               name="org.opencastproject.adminui.impl.MediaPackageLockService"
+               immediate="true">
+  <implementation class="org.opencastproject.adminui.impl.MediaPackageLockService" />
+  <property name="service.description" value="MediaPackage Lock Service" />
+  <service>
+	<provide interface="org.osgi.service.cm.ManagedService" />
```
Comment (lkiesow) [inline modules/admin-ui/src/main/resources/OSGI-INF/mediapackage_lock.xml]: purely cosmetic: indentation ;-P

**Comment by smarquard** [comment]
It seems much more helpful to me to show that another user has the editor open than to actually lock it. Two people could lock at the same video at the same time.

Also re-editing a video after publication is a perfectly reasonable use-case (something we do here quite often).

Lastly, Harvard has had some code for a while that is a generalized lock service. May be worth pulling that in rather than the more specific mediapackage lock service here?

**Comment by ts23** [comment]
The current editor AFAIK does not allow competitive editing, so keeping it open with a message somebody else is editing still does not stop the potential data corruption.

Reediting after publication caused some failing workflows in our case that is why it was taken out here 
our workflow needs the publication to be retracted and then resubmitted. I can take this part out of the PR if that is a preferred choice. The lock of reediting after publish might not be an issue as it is checking if the there are previews available. It only locks if there are none ... so it might be an unfortunate message rather than a stopping of reedits.

**Comment by smarquard** [comment]
What is the "potential data corruption"? i.e. is it just that the last person to save has their edits winning, or some other issue?

For the re-editing after publication, that seems like something that should be configurable but also then if a specific mediapackage wasn't re-editable then that should also be indicated in the Admin UI somehow, i.e. the edit option is no longer available or greyed out?

It seems right that the editor should show an error message if there are no previews available.

**Comment by ts23** [comment]
I think it is just the last save wins scenario

This is now implemented, There are two parameters to define if the editor locks or just informs.
It also provides info who is editing.

**Comment by karendolan** [comment]
@ts23 & @smarquard A generalized mediapackage lock service would be very helpful. Our site has several processes that start workflows behind the scenes (transcriptions from 3rd party systems, metadata updates from 3rd party systems, asset uploads from archived events via the admin-ui, capture agents, etc). Once a process starts its workflow, its all Ok, because the workflow creates a sort of lock. But, there is a large dangerous race condition between the time a process starts to save/upload data and files and the time it starts the workflow. A mediapackage Lock Managment Service would bridge that gap and also protect the UI editor (as a service used by the admin UI). 

@ts23 I'll discuss this with @rute-santos (who wrote our Mediapackage Lock Management module for our production v1.63). If this sounds Ok to you, we can make a complimentary patch.

**Comment by karendolan** [comment]
Hi @ts23, I talked to @rute-santos , who thinks our implementation of mp locking has to be more generalized for the community and we won't be able to look at 5x mp locking scenarios in depth for a while. So, I'm stepping away from the conversation for now.

**Comment by rute-santos** [comment]
Hi @smarquard, @ts23, @karendolan,
Our media package lock service had the limitation of only locking among services running on the same node e.g. admin, which worked fine for us. I believe this could be solved by using the database to store the lock (and pessimistic transactions).
We don't intend to use it anymore but, if you want the code, you can find it at https://bitbucket.org/opencast-community/matterhorn-harvard-fork/pull-requests/11.

File: modules/admin-ui/package-lock.json [hunk 2]
```diff
@@ -0,0 +1,9920 @@
+{
```
Comment (wsmirnow) [inline modules/admin-ui/package-lock.json:1]: The package-lock should not be part of this pull request. Please remove it.

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/ToolsEndpoint.java [hunk 3]
```diff
@@ -774,11 +840,13 @@ public int compare(Tuple<Long, Long> t1, Tuple<Long, Long> t2) {
     private final List<Tuple<Long, Long>> segments;
     private final List<String> tracks;
     private final Opt<String> workflow;
+    private final Opt<String> autosave;
```
Comment (wsmirnow) [inline modules/admin-ui/src/main/java/org/opencastproject/adminui/endpoint/ToolsEndpoint.java]: Why is the autosave option a string and not a boolean?

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/MediaPackageLockService.java [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+
+  public MediaPackageLockService() {
+    super();
+    logger.info("starting MP Lock service");
+  }
+
+  public long getMediaPackageLock(Event event, String sessionId) {
+    EditorLock releaseLock = null;
+    EditorLock lock = new EditorLock(event.getIdentifier(), sessionId);
+    logger.debug("lock mp {} for session {}",event.getIdentifier(), sessionId);
```
Comment (wsmirnow) [inline modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/MediaPackageLockService.java]: please add a whitespace after ','

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/MediaPackageLockService.java [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        lock.setDuration(lockDur);
+        mpLock.put(lock.getMediaPackageId(), lock);
+        return -1;
+      }
+      return releaseLock.checkLock(lock);
+    }
+  }
+
+  public void releaseMediaPackageLock(final Event event, String sessionId) {
+    logger.debug("unlock mp {}",event.getIdentifier());
```
Comment (wsmirnow) [inline modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/MediaPackageLockService.java]: please add a whitespace after ','

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/MediaPackageLockService.java [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+   * The Json key for the SessionId.
+   */
+  public static final String SESSION_KEY = "session";
+
+  public MediaPackageLockService() {
+    super();
+    logger.info("starting MP Lock service");
+  }
+
+  public long getMediaPackageLock(Event event, String sessionId) {
```
Comment (wsmirnow) [inline modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/MediaPackageLockService.java]: documentation needed

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/MediaPackageLockService.java [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+        }
+        lock.setDuration(lockDur);
+        mpLock.put(lock.getMediaPackageId(), lock);
+        return -1;
+      }
+      return releaseLock.checkLock(lock);
+    }
+  }
+
+  public void releaseMediaPackageLock(final Event event, String sessionId) {
```
Comment (wsmirnow) [inline modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/MediaPackageLockService.java]: please add documentation

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/MediaPackageLockService.java [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+      }
+    }
+  }
+
+  @Override
+  public void updated(Dictionary<String, ?> dctnr) throws ConfigurationException {
+    throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
+  }
+
+  public void cleanUp() {
```
Comment (wsmirnow) [inline modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/MediaPackageLockService.java]: documentation

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/MediaPackageLockService.java [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+        if (el.removable(sessionId)) {
+          mpLock.remove(event.getIdentifier());
+        }
+      }
+    }
+  }
+
+  @Override
+  public void updated(Dictionary<String, ?> dctnr) throws ConfigurationException {
+    throw new UnsupportedOperationException("Not supported yet."); //To change body of generated methods, choose Tools | Templates.
```
Comment (wsmirnow) [inline modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/MediaPackageLockService.java]: Throwing an exception in a OSGI method should be avoid. If you are not using the `updated(Dictionary)` method, you should drop `ManagedService` implementation

**Review by wsmirnow — CHANGES_REQUESTED** [review]
First part of the review. WIP

File: modules/admin-ui/src/main/webapp/scripts/modules/events/subresources/controllers/toolsController.js [hunk 10]
```diff
@@ -73,17 +73,41 @@ angular.module('adminNg.controllers')
...
+        $scope.release = function() {
+          ToolsResource.release({id: $scope.id, tool: 'lock'});
+        };
         $scope.submit = function () {
+            $scope.video.autosave = false;
             $scope.submitButton = true;
             $scope.video.$save({ id: $scope.id, tool: $scope.tab }, function () {
                 $scope.submitButton = false;
                 if ($scope.video.workflow) {
+                    $scope.video.autosave = false;
```
Comment (wsmirnow) [inline modules/admin-ui/src/main/webapp/scripts/modules/events/subresources/controllers/toolsController.js]: this line can be removed as the value isn't changed

File: modules/admin-ui/src/main/webapp/scripts/modules/events/subresources/controllers/toolsController.js [hunk 11]
```diff
@@ -73,17 +73,41 @@ angular.module('adminNg.controllers')
...
+          if ($scope.video.status === 'locked' ) {
+            var mins = $scope.video.locked;
+            Notifications.addWithParams('error', 'VIDEO_EDIT_LOCKED_MINS', {minutes : mins});
+            $location.url('/events/' + $scope.resource);
+          }
+          if ($scope.video.status === 'edited before' ) {
+            Notifications.add('error', 'VIDEO_EDITED_BEFORE', 'video-tools');
+            $location.url('/events/' + $scope.resource);
+          }
+        });
```
Comment (wsmirnow) [inline modules/admin-ui/src/main/webapp/scripts/modules/events/subresources/controllers/toolsController.js]: The 404 isn't handled yet. As it wasn't handled before, we can create an other ticket for fixing it.

**Review by wsmirnow — CHANGES_REQUESTED** [review]

File: modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/MediaPackageLockService.java (hunk 9 above)
Comment (ts23) [reply modules/admin-ui/src/main/java/org/opencastproject/adminui/impl/MediaPackageLockService.java]: I need managed service otherwise I cannot register the service through OSGI

File: modules/admin-ui/src/main/webapp/scripts/modules/events/subresources/controllers/toolsController.js (hunk 11 above)
Comment (ts23) [reply modules/admin-ui/src/main/webapp/scripts/modules/events/subresources/controllers/toolsController.js]: I can fix this in here

**Comment by JamesUoM** [comment]
I've found a few issues with the code itself the and it also appears that part of our autosave facility has been committed in this PR. I've closed this PR and I will create a new one based on our latest version.
